Disable LDL/STL checks for CTK < 13.1 (nvbug 5243118)

lmcrean opened the pull request:
## Problem statement

Issue #5560: c.parallel tests incorrectly perform SASS validation for LDL/STL instructions on CTK < 13.1, causing false positives due to nvbug 5243118 where nvrtc generates these instructions when nvcc would not.

## Solution

Since c.parallel **always** uses nvrtc internally for kernel compilation (not nvcc), the fix unconditionally applies CTK version checks:

### C++ Changes
- Added `is_ctk_version_allows_sass_check()` helper function in `test_util.h` that checks CTK version using `__CUDACC_VER_MAJOR__` and `__CUDACC_VER_MINOR__` macros
- Returns `true` for CTK ≥ 13.1 (where bug is fixed), `false` otherwise
- Updated `should_check_sass()` in `test_scan.cpp` and `test_unique_by_key.cpp` to use this helper

### Python Changes  
- Added `_should_check_sass_for_ctk_version()` in `_cccl_interop.py` that checks runtime CTK version
- Modified `call_build()` to only perform SASS checks when both `_check_sass` is enabled AND CTK ≥ 13.1
- Updated `conftest.py` fixture to emit warning when SASS checks are disabled due to CTK version
- Simplified `test_scan.py` to remove CC 9.0+ check (now handled by CTK version check)

## Testing guide

**Expected behavior:**
- CTK < 13.1: SASS validation disabled (with warning), tests pass without false positives
- CTK ≥ 13.1: SASS validation enabled, properly catches real LDL/STL issues
- Complex types still skip SASS checks via existing logic

## Breaking Changes

**None** - This is a backward-compatible fix that only affects internal test validation logic. No changes to public APIs, library functionality, build system, or dependencies.

## Checklist
- [x] I am familiar with the [Contributing Guidelines]().
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

miscco reviewed (changes requested):
Thanks a lot for working on this.

I have some concerns about disabling the checks for all NVIDIA compilers

lmcrean commented on c/parallel/test/test_scan.cpp:
```diff
@@ -74,8 +74,17 @@ struct scan_build
+    bool cc_allows_check = (!Disable75SassCheck || cc_major > 7) && cc_major < 9;
+
+    // Check CTK version for nvrtc LDL/STL bug (resolves nvbug 5243118)
+#if defined(__CUDACC_VER_MAJOR__) && defined(__CUDACC_VER_MINOR__)
```
Thank you for the review, I have tried to address this

miscco commented:
/ok to test 035251c

rwgk commented on c/parallel/test/test_scan.cpp:
```diff
@@ -74,8 +74,25 @@ struct scan_build
 
   static bool should_check_sass(int cc_major)
```
Did you consider moving this code to `test_util.h`? Avoiding duplication of this non-trivial code will help long-term maintenance.

There is a member function with the same name in `algorithm_execution.h`. I'm guessing that's fine as-is, but I'm pointing it out JIC.

lmcrean replied:
Thanks for the review -- I thought about it and moved the duplicated nvrtc version check logic to a shared helper function as suggested:

`test_util.h` now uses:
```h
inline bool is_nvrtc_sass_check_allowed()
{
#ifdef __CUDACC_RTC__
    // We're compiling with nvrtc - check CTK version for the bug
#  if defined(__CUDACC_VER_MAJOR__) && defined(__CUDACC_VER_MINOR__)
    // Bug is fixed in CTK 13.1+
    return (__CUDACC_VER_MAJOR__ > 13) || (__CUDACC_VER_MAJOR__ == 13 && __CUDACC_VER_MINOR__ >= 1);
#  else
    // Conservative: disable if version unknown
    return false;
#  endif
#else
    // We're compiling with nvcc - no CTK version restriction needed
    // nvbug 5243118 does not affect nvcc compilation
    return true;
#endif
}
```

lmcrean replied:
Regarding `should_check_sass` in `algorithm_execution.h`, my understanding is that it serves a different purpose there:

1. `should_check_sass` in **`algorithm_execution.h`** provides the template framework with a default "always check" behavior via `build_traits`
2. `should_check_sass` in **individual test files** (`test_scan.cpp`, `test_unique_by_key.cpp`) override this default with their algorithm-specific compute capability requirements

lmcrean committed:
Disable LDL/STL checks for CTK < 13.1 (nvbug 5243118)

rwgk reviewed (approved):
Nice, thanks!

lmcrean commented:
@miscco could we run CI please?

NaderAlAwar reviewed (changes requested):
Hi @lmcrean, thanks for the PR! This currently does not address the main issue. In c.parallel, we use nvrtc to compile template instantiations of the kernels. You can find examples of this in files under `c/parallel/src`, such as `scan.cu`. Depending on the nvrtc version (anything before CTK 13.1), the compiled code might include LDL and STL instructions due to a bug in nvrtc. We know this is a bug because compiling the same kernel template instantiation using nvcc (which is what we do in C++ CUB) does not yield those problematic instructions.

The issue with your current approach is that it uses a macro to check if the test code is compiled with nvrtc. However, we always compile the test code with nvcc. The nvrtc compilation is happening internally (in the `c/parallel/src/` files I mentioned above). Therefore, some changes are required:

1. We don't need to check if we are compiling with nvrtc. c.parallel always uses nvrtc internally
2. The logic should be changed so that we never check the SASS for CTK versions older than 13.1.

Additionally, we want to add a similar check on the python side as well. Let me know if you need more context there. Thanks!

lmcrean committed:
Refactor SASS checks for CTK version compatibility

- Updated logic to disable SASS checks for CTK versions below 13.1 due to nvrtc bug (nvbug 5243118) in multiple test files.
- Introduced a new function to check CTK version compatibility for SASS checks.
- Adjusted test fixtures to conditionally enable SASS verification based on CTK version.

lmcrean committed:
Enhance CUDA runtime version detection in SASS checks

- Improved the logic for detecting the CUDA runtime version by implementing multiple fallback methods, including checks for the cuda-python package and numba's CUDA detection.
- Added warnings for cases where the CUDA version cannot be detected, ensuring SASS verification is disabled appropriately.
- This update aims to provide more robust handling of CUDA version checks for SASS compatibility.

lmcrean commented:
@NaderAlAwar 

Thanks for the detailed feedback and clarification! Appreciate it. I was initially approaching this incorrectly by checking nvrtc compilation when the real issue is the CTK version dependency.

I've updated the PR to address your points:
- Replaced the TODO comments about adding NVRTC version checks with proper CTK version logic that recognizes c.parallel always uses nvrtc internally
- Added `ctk_version_allows_sass_check()` in `test_util.h` that uses `__CUDACC_VER_MAJOR__` and `__CUDACC_VER_MINOR__` to disable SASS checks for CTK < 13.1, directly addressing nvbug 5243118
- Added `_should_check_sass_for_ctk_version()` in `_cccl_interop.py` with runtime CTK version detection. The `call_build()` function now only performs SASS checks when both `_check_sass` is enabled AND CTK ≥ 13.1

The tests are now comprehensive. This should prevent the false positives from nvrtc's LDL/STL instruction generation while maintaining proper validation for CTK 13.1+. Let me know if you'd like me to adjust anything about this approach

lmcrean commented:
@miscco can we run /ok to test please?

rwgk commented:
/ok to test cbe2ef0

lmcrean committed:
Streamline CUDA runtime version detection for SASS checks

- Replaced multiple fallback methods for detecting the CUDA runtime version with a direct call to `cudart.cudaRuntimeGetVersion()`.
- Improved error handling to provide clearer warnings when version detection fails, ensuring SASS verification is disabled appropriately.
- This update enhances the reliability and simplicity of CUDA version checks for SASS compatibility.

lmcrean commented:
Can we check CI please?

lmcrean commented:
@shwina thanks for keeping me in the loop, this makes sense, look forward to getting it over the final hurdle when ready
